### PR TITLE
Allow AQL variable names starting with an underscore, as stated in the docs

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Docs PR: https://github.com/arangodb/docs/pull/689

Allow AQL variable names to start with an underscore character, as is stated in the docs.
The AQL parser actually always allowed this, but there was an additional check inside the variable handling code that actively disallowed such variable names.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [x] No backports required

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This PR adds tests that were used to verify all changes:
  - [x] Added new **integration tests** (shell_server_aql)

### CODE REVIEW

**Comment by 153957** [comment]
Fixes https://github.com/arangodb/arangodb/issues/13513

**Review by Simran-B — APPROVED** [review]
I suppose this was meant to reserve variable names with underscores for internal purposes?
Existing pseudo-variables don't make use of that though (OLD, NEW) and there is also namespacing which could be utilized, but I don't currently see a need for that.

**Comment by 153957** [comment]
Should a test be added to indicate that `_` is an invalid variable? (Also anything matching the regex pattern `_[^a-zA-Z0-9].*` should be invalid)

Since that is stated in the docs:
> If a variable name starts with the underscore character, the underscore must be followed by least one letter (a-z or A-Z) or digit (0-9).

**Comment by Simran-B** [comment]
A single underscore can be used as a variable name if it's escaped: `` LET `_` = 1 RETURN `_` ``
However, I think we should only describe what is allowed without escaping as variable names in the documentation.

What's interesting is that in a query explain where variables are usually referenced like `#1`, `#2` etc., it shows `#_` with the above query:

```
Execution plan:
 Id   NodeType          Est.   Comment
  1   SingletonNode        1   * ROOT
  2   CalculationNode      1     - LET #_ = 1   /* json expression */   /* const assignment */
  4   ReturnNode           1     - RETURN #_
```

BTW. Escaping a single digit like `` LET `1` = 1 RETURN `1` `` hits this assertion:
https://github.com/arangodb/arangodb/blob/feature/allow-variable-names-starting-with-underscore/arangod/Aql/VariableGenerator.cpp#L68

```
2021-04-20T14:19:35Z [3635] FATAL [a7902] {crash} 💥 ArangoDB 3.9.0-devel enterprise [linux], thread 4 [SchedWorker] caught unexpected signal 6 (SIGABRT): assertion failed in /work/ArangoDB/arangod/Aql/VariableGenerator.cpp:68 [createVariable]: !isUserDefined || variable->isUserDefined() - image base address: 0x0000000000400000
2021-04-20T14:19:35Z [3635] INFO [c962b] {crash} Backtrace of thread 4 [SchedWorker]
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  1 [+0x0000000001d40857] arangodb::CrashHandler::crash(char const*) (+0x17)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  2 [+0x0000000001d40a05] arangodb::CrashHandler::assertionFailure(char const*, int, char const*, char const*) (+0xf5)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  3 [+0x00000000007d88b3] arangodb::aql::VariableGenerator::createVariable(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool) (+0x173)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  4 [+0x000000000080ff85] arangodb::aql::Ast::createNodeVariable(char const*, unsigned long, bool) (+0x1b5)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  5 [+0x000000000081014a] arangodb::aql::Ast::createNodeLet(char const*, unsigned long, arangodb::aql::AstNode const*, bool) (+0x5a)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  6 [+0x0000000000bbf98a] Aqlparse(arangodb::aql::Parser*) (+0x1daa)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  7 [+0x0000000000b1e764] arangodb::aql::Parser::parse() (+0x84)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  8 [+0x0000000000794fb7] arangodb::aql::Query::preparePlan() (+0x1b7)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame  9 [+0x000000000079578f] arangodb::aql::Query::prepareQuery(arangodb::aql::SerializationFormat) (+0x5f)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 10 [+0x000000000079b905] arangodb::aql::Query::execute(arangodb::aql::QueryResult&) (+0x9d5)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 11 [+0x00000000004909fc] arangodb::RestCursorHandler::processQuery() (+0x2c)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 12 [+0x000000000049772e] arangodb::RestCursorHandler::registerQueryOrCursor(arangodb::velocypack::Slice const&) (+0x66e)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 13 [+0x0000000000497d1a] arangodb::RestCursorHandler::createQueryCursor() (+0x9a)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 14 [+0x000000000042d81d] arangodb::rest::RestHandler::executeEngine(bool) (+0xcd)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 15 [+0x000000000042ed0a] arangodb::rest::RestHandler::runHandlerStateMachine() (+0x40a)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 16 [+0x0000000000726aac] arangodb::Scheduler::WorkItem<arangodb::rest::CommTask::handleRequestAsync(std::shared_ptr<arangodb::rest::RestHandler>, unsigned long*)::{lambda()#1}>::invoke() (+0xdc)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 17 [+0x0000000000560e58] arangodb::SupervisedScheduler::runWorker() (+0x1e8)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 18 [+0x0000000001d5f63c] arangodb::Thread::runMe() (+0x1c)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 19 [+0x0000000001d5f6fe] arangodb::Thread::startThread(void*) (+0x6e)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 20 [+0x0000000001e12de6] ThreadStarter(void*) (+0x36)
2021-04-20T14:19:35Z [3635] INFO [308c3] {crash} frame 21 [+0x000000000379d7a3] start (+0x70)
2021-04-20T14:19:35Z [3635] INFO [ded81] {crash} available physical memory: 67513589760, rss usage: 198770688, vsz usage: 1248014336, threads: 55
fish: './build/bin/arangod --server.en…' terminated by signal SIGABRT (Abort)
```

**Comment by jsteemann** [comment]
@Simran-B : reproduced...
I just tried adding an additional validation of the variable name via a C++ `std::regex` in the code, only to find out that a regex such as `^(\\$?[a-zA-Z][_a-zA-Z0-9]*|_+[a-zA-Z]+[_a-zA-Z0-9]*)$`, even when simplified to `^(\\$?|_+)[a-zA-Z]+[_a-zA-Z0-9]*$`, uses up to 25,000 stack frames and then crashes the program for an input string such as `variablevariablevariablevariable...` (`variable` repeated 10,000).
Hand-rolling a validation...

**Comment by goedderz** [comment]
The `std::regex` implementation seems to have some problems. For long strings like `"variable" x 10000`, even `.*` or `^.*$` crash with a stack overflow. I've tried some other implementations (perl, libpcre, python, GNU grep, ripgrep, nodejs), and all of them could handle those regexes fine, even after I've increases the string's size by another factor 1000.

Looks to me like we should avoid `std::regex`.

**Comment by jsteemann** [comment]
@goedderz : fully agree. Looks like `std::regex` is powered by satan.

File: arangod/Aql/VariableGenerator.cpp [hunk 1]
```diff
@@ -168,6 +172,36 @@ void VariableGenerator::fromVelocyPack(VPackSlice const slice) {
     createVariable(var);
   }
 }
+
+bool VariableGenerator::isValidName(char const* p, char const* end) noexcept {
```
Comment (gnusi) [inline arangod/Aql/VariableGenerator.cpp:176]: why not https://compile-time-regular-expressions.readthedocs.io/en/latest/?

**Comment by gnusi** [comment]
@jsteemann there are no tests :cry:

**Comment by jsteemann** [comment]
@gnusi: there are tests: https://github.com/arangodb/arangodb/pull/13745/files#diff-758ea50fd18e8df9dcc32b98ceae715c5b8df6d1ac192f8d9b7ea6b644e240b4

**Comment by gnusi** [comment]
@jsteemann indeed, sorry I overlooked it

**Pull request merged**